### PR TITLE
SNOW-877039 Ingest SDK retry for INVALID JWT Response

### PR DESCRIPTION
**Pull request by sfc-gh-alhuang**
According to [SNOW-877039](https://snowflakecomputing.atlassian.net/browse/SNOW-877039) and [SNOW-876912](https://snowflakecomputing.atlassian.net/browse/SNOW-876912) JWT might expire at some point even if the token is refreshed 5 minutes earlier. Add a retry mechanism to resolve this issue.

[SNOW-877039]: https://snowflakecomputing.atlassian.net/browse/SNOW-877039?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[SNOW-876912]: https://snowflakecomputing.atlassian.net/browse/SNOW-876912?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: src/main/java/net/snowflake/ingest/connection/SecurityManager.java
Comment (sfc-gh-japatel) [inline src/main/java/net/snowflake/ingest/connection/SecurityManager.java:93]: nit: Javadocs please.

File: src/main/java/net/snowflake/ingest/connection/OAuthManager.java
Comment (sfc-gh-japatel) [inline src/main/java/net/snowflake/ingest/connection/OAuthManager.java:14]: lets keep this class final.

File: src/main/java/net/snowflake/ingest/connection/JWTManager.java
Comment (sfc-gh-japatel) [inline src/main/java/net/snowflake/ingest/connection/JWTManager.java:25]: lets also keep this class final

File: src/main/java/net/snowflake/ingest/utils/Constants.java
Comment (sfc-gh-japatel) [inline src/main/java/net/snowflake/ingest/utils/Constants.java:57]: `..FOR_UNAUTHORIZED_RETRY`

File: src/main/java/net/snowflake/ingest/streaming/internal/StreamingIngestUtils.java
Comment (sfc-gh-japatel) [inline src/main/java/net/snowflake/ingest/streaming/internal/StreamingIngestUtils.java:139]: Lets add a logger info here

Comment (sfc-gh-japatel) [reply src/main/java/net/snowflake/ingest/streaming/internal/StreamingIngestUtils.java:139]: Can we implement a retry in `handleExceptionalStatus`? that will benefit other apis as well (snowpipe)

File: src/main/java/net/snowflake/ingest/connection/SecurityManager.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/connection/SecurityManager.java:93]: Added.

File: src/main/java/net/snowflake/ingest/streaming/internal/StreamingIngestUtils.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/streaming/internal/StreamingIngestUtils.java]: Sounds good, moved into `handleExceptionalStatus`

File: src/main/java/net/snowflake/ingest/connection/JWTManager.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/connection/JWTManager.java:25]: Remove final since `mockito.spy()` is used to simulate the case encounter invalid token. There might a [solution](https://stackoverflow.com/questions/14292863/how-to-mock-a-final-class-with-mockito) which required updating some dependencies. Do you think we should keep final and do this?

File: src/main/java/net/snowflake/ingest/utils/Constants.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/utils/Constants.java]: Resolved.

File: src/main/java/net/snowflake/ingest/connection/JWTManager.java
Comment (sfc-gh-japatel) [reply src/main/java/net/snowflake/ingest/connection/JWTManager.java:25]: ahh I see, missed the nuances of final and mockito. 

I do see powermock in our dependency, can you try using that? 

This is a security class and I would like to avoid anyone even remotely trying to extend this class. Let me know if you have any issues, we can chat offline.

File: src/main/java/net/snowflake/ingest/connection/ServiceResponseHandler.java
Comment (sfc-gh-japatel) [inline src/main/java/net/snowflake/ingest/connection/ServiceResponseHandler.java:267]: Lets keep this WARN, instead of info
`"Authorization failed, refreshing Token succeeded, retries={}"  `

File: src/test/java/net/snowflake/ingest/streaming/internal/StreamingIngestUtilsTest.java
Comment (sfc-gh-japatel) [inline src/test/java/net/snowflake/ingest/streaming/internal/StreamingIngestUtilsTest.java:182]: can you also add a test which makes sure it doesnt go in an infinite loop? for instance max it should retry is 3 and after that it should go in securityException

File: src/main/java/net/snowflake/ingest/connection/ServiceResponseHandler.java
Comment (sfc-gh-japatel) [inline src/main/java/net/snowflake/ingest/connection/ServiceResponseHandler.java:248]: is it going in an infinite loop if exception is non SC_UNAUTHORIZED?

File: src/main/java/net/snowflake/ingest/connection/ServiceResponseHandler.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/connection/ServiceResponseHandler.java:267]: Fixed.

File: src/main/java/net/snowflake/ingest/connection/ServiceResponseHandler.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/connection/ServiceResponseHandler.java:248]: I think it would throw exception immediately if status is neither 401 nor 503. ([ref](https://github.com/snowflakedb/snowflake-ingest-java/blob/80caea6f0fc85177a374bcd1cd281a8c241559e9/src/main/java/net/snowflake/ingest/connection/ServiceResponseHandler.java#L273))

File: src/test/java/net/snowflake/ingest/streaming/internal/StreamingIngestUtilsTest.java
Comment (sfc-gh-alhuang) [reply src/test/java/net/snowflake/ingest/streaming/internal/StreamingIngestUtilsTest.java:214]: Added [here](https://github.com/snowflakedb/snowflake-ingest-java/blob/6927aaaf0b68cfa6c1e40df28512a535d0f6a661/src/test/java/net/snowflake/ingest/streaming/internal/StreamingIngestUtilsTest.java#L164) and [here](https://github.com/snowflakedb/snowflake-ingest-java/blob/6927aaaf0b68cfa6c1e40df28512a535d0f6a661/src/test/java/net/snowflake/ingest/streaming/internal/StreamingIngestUtilsTest.java#L184).

File: src/main/java/net/snowflake/ingest/connection/JWTManager.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/connection/JWTManager.java]: I've achieved it through PowerMockito, thanks for the tips!

Comment (sfc-gh-japatel) [reply src/main/java/net/snowflake/ingest/connection/JWTManager.java]: Good stuff! Thank you!

File: src/main/java/net/snowflake/ingest/connection/ServiceResponseHandler.java
Comment (sfc-gh-japatel) [reply src/main/java/net/snowflake/ingest/connection/ServiceResponseHandler.java:248]: ack!

File: src/test/java/net/snowflake/ingest/streaming/internal/StreamingIngestUtilsTest.java
Comment (sfc-gh-japatel) [inline src/test/java/net/snowflake/ingest/streaming/internal/StreamingIngestUtilsTest.java:179]: awesome, good to see these tests, thanks for adding them!

**Review by sfc-gh-japatel — APPROVED** [review]
Thanks Alec, lgtm. lets get an approval from @sfc-gh-tzhang as well and lets wait for all merge gates to pass before we push it!

File: src/main/java/net/snowflake/ingest/utils/Constants.java
Comment (sfc-gh-tzhang) [inline src/main/java/net/snowflake/ingest/utils/Constants.java:57]: I don't understand why we need a retry on this, if it works with one refresh then it will work, otherwise it will never work no matter how many times you retry, right?

File: src/main/java/net/snowflake/ingest/SimpleIngestManager.java
Comment (sfc-gh-tzhang) [inline src/main/java/net/snowflake/ingest/SimpleIngestManager.java:594]: For my own reference, why do we need to move this out?

File: src/main/java/net/snowflake/ingest/connection/RequestBuilder.java
Comment (sfc-gh-tzhang) [inline src/main/java/net/snowflake/ingest/connection/RequestBuilder.java:570]: Trying to understand why the code was written this way before, can this token be updated in securityManager after it's passed into addToken?

File: src/main/java/net/snowflake/ingest/connection/ServiceResponseHandler.java
Comment (sfc-gh-tzhang) [inline src/main/java/net/snowflake/ingest/connection/ServiceResponseHandler.java:180]: do we use the response anywhere?

File: src/main/java/net/snowflake/ingest/connection/ServiceResponseHandler.java
Comment (sfc-gh-tzhang) [inline src/main/java/net/snowflake/ingest/connection/ServiceResponseHandler.java:263]: Duplicate logic, could you log it before the all cases?

File: src/main/java/net/snowflake/ingest/utils/Constants.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/utils/Constants.java]: Make sense, removed the mutiple retries logic

File: src/main/java/net/snowflake/ingest/connection/ServiceResponseHandler.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/connection/ServiceResponseHandler.java:179]: [Here](https://github.com/snowflakedb/snowflake-ingest-java/blob/6927aaaf0b68cfa6c1e40df28512a535d0f6a661/src/main/java/net/snowflake/ingest/connection/ServiceResponseHandler.java#L184)

File: src/main/java/net/snowflake/ingest/connection/ServiceResponseHandler.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/connection/ServiceResponseHandler.java]: Fixed.

File: src/main/java/net/snowflake/ingest/SimpleIngestManager.java
Comment (sfc-gh-alhuang) [reply src/main/java/net/snowflake/ingest/SimpleIngestManager.java:594]: The request would be used twice later at [here](https://github.com/snowflakedb/snowflake-ingest-java/blob/6927aaaf0b68cfa6c1e40df28512a535d0f6a661/src/main/java/net/snowflake/ingest/SimpleIngestManager.java#L596) and [here](https://github.com/snowflakedb/snowflake-ingest-java/blob/6927aaaf0b68cfa6c1e40df28512a535d0f6a661/src/main/java/net/snowflake/ingest/SimpleIngestManager.java#L599). Moved it out side to avoid building it twice.

**Review by sfc-gh-tzhang — APPROVED** [review]
lgtm, thanks!

File: src/main/java/net/snowflake/ingest/connection/ServiceResponseHandler.java
Comment (sfc-gh-tzhang) [inline src/main/java/net/snowflake/ingest/connection/ServiceResponseHandler.java:238]: It's kind of weird that we're passing HttpResponse as an input and return a HttpResponse as well, a better way to do it is to retry the request before calling this function and pass the final the final response here, this is just a suggestion and you don't have to change it if it involves too many changes

**Pull request merged**
